llpc/lower/llpcSpirvLowerRayTracingIntrinsics.cpp

```diff
@@ -183,4 +211,108 @@ void SpirvLowerRayTracingIntrinsics::createConvertF32toF16(Function *func, unsig
   m_builder->CreateRet(result);
 }
 
+// =====================================================================================================================
+// Create AmdExtGetBaseAddrFromResource
+//
+// @param func : Function to create
+void SpirvLowerRayTracingIntrinsics::createGetBaseAddrFromResource(Function *func) {
+  // uint64_t AmdExtGetBaseAddrFromResource(resource)
+  // {
+  //   return resource.baseAddr;
+  // }
+
+  assert(func->getBasicBlockList().size() == 1);
+  (*func->begin()).eraseFromParent();
+
+  BasicBlock *entryBlock = BasicBlock::Create(m_builder->getContext(), "", func);
+  m_builder->SetInsertPoint(entryBlock);
+  auto argIt = func->arg_begin();
+
+  Value *zero = m_builder->getInt32(0);
+
+  Value *resourcePtr = m_builder->CreateLoad(argIt->getType()->getPointerElementType(), argIt);
+  Value *ptr = m_builder->CreateGEP(resourcePtr->getType()->getPointerElementType(), resourcePtr, {zero});
+
+  m_builder->CreateRet(m_builder->CreatePtrToInt(ptr, m_builder->getInt64Ty()));
+}
+
+// =====================================================================================================================
+// Create AmdExtAtomicFMinAtAddrx2, AmdExtAtomicFMinAtAddr, AmdExtAtomicFMaxAtAddrx2, AmdExtAtomicFMaxAtAddr
+//
+// @param func : Function to create
+// @param isMin : Whether is Min operation, otherwise Max operation
+// @param is64Ty : Whether is 64-bit type operation, otherwise 32-bit type operation
+void SpirvLowerRayTracingIntrinsics::createAtomicFMinMaxAtAddr(Function *func, bool isMin, bool is64Ty) {
+  // float(2) AmdExtAtomicFMinAtAddr(x2)(uint64_t gpuVa, uint offset, float(2) value)
+
+  assert(func->getBasicBlockList().size() == 1);
+  (*func->begin()).eraseFromParent();
+
+  BasicBlock *entryBlock = BasicBlock::Create(m_builder->getContext(), "", func);
+  m_builder->SetInsertPoint(entryBlock);
+  auto argIt = func->arg_begin();
+
+  Type *gpuVaTy = m_builder->getInt64Ty();
+  Type *valueTy = is64Ty ? cast<Type>(FixedVectorType::get(m_builder->getFloatTy(), 2)) : m_builder->getFloatTy();
+  // TODO: Remove this when LLPC will switch fully to opaque pointers.
+  assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(argIt->getType(), gpuVaTy));
+  Value *gpuVa = m_builder->CreateLoad(gpuVaTy, argIt++);
+  Value *offset = m_builder->CreateLoad(m_builder->getInt32Ty(), argIt++);
```

Comment:
Using `func->getArg(i)` instead of an incrementing iterator is cleaner. Please do that instead, also in the other function.